routes/admin.js

```diff
@@ -0,0 +1,280 @@
+    }
+  });
+
+  /* API method to get database space in use */
+  app.get("/admin/dashboard/pg-space", async (req, res, next) => {
+    try {
+
+      let log = null;
+
+      await pool.query('select get_database_size();', async (err, result) => {
+        if (err) {
+          console.log(err)
+          return;
+        }
+        log = result.rows;
+        res.json(log);
+      });
+
+    } catch (e) {
+      return next(e);
+    }
+  });
+
+  /* Login */
+  app.get("/admin/login", userIsNotAuthenticated, (req, res) => {
+    res.render("login.ejs", { message: null });
+  });
+
+  /* Register new user (NOTE: this is an admin privilege only, and is intentionally *not* outward facing) */
+  app.get('/admin/register', userIsAdmin, (req, res) => {
+    res.render('registerUser.ejs');
+  });
+
+  app.post('/admin/register', userIsAdmin, (req, res) => {
+    const registerUser = async () => {
+      try {
+        const { name, role, email, password } = req.body;
+        const hashedPassword = await bcrypt.hash(password, 10);
+        const newUser = await pool.query('INSERT INTO production_user (name, role, email, password) VALUES ($1, $2, $3, $4)', [name, role, email, hashedPassword]
+        );
+        console.log(`You have succesfully registered user ${email}.`);
+        res.render('dashboard');
+      } catch (err) {
+        console.log(err);
+      }
+    }
+    registerUser();
+  });
+
+  /* Change password */
+  /* TODO: add logic to make this work, so that if a user has a default password they are required to change it */
+  app.get('/admin/changePassword', userIsAuthenticated, (req, res) => {
+    res.render('changePassword.ejs');
+  });
+
+  /* Logout */
+  app.get("/admin/logout", (req, res) => {
+    req.logout();
+    res.render("login.ejs", { message: "You have logged out successfully" });
+  });
+
+  /* Handle input from the login form */
+  app.post(
+    "/admin/login",
+    passport.authenticate("local", {
+      successRedirect: "/admin/dashboard",
+      failureRedirect: "/admin/login",
+      failureFlash: true
+    })
+  );
+
+  /* Passport middleware function to protect routes */
+  function userIsNotAuthenticated(req, res, next) {
+    if (req.isAuthenticated()) {
+      return res.redirect("/admin/dashboard");
+    }
+    next();
+  }
+
+  /* Passport middleware function to protect routes */
+  function userIsAuthenticated(req, res, next) {
+    if (req.isAuthenticated()) {
+      return next();
+    }
+    res.redirect("/admin/login");
+  }
+
+  //this is for the "Create User" route, which should be accesible by logged-in Admin users only
+  function userIsAdmin(req, res, next) {
+    if (req.isAuthenticated() && req.user.role === "admin") {
+      return next();
+    }
+    res.redirect("/admin/dashboard");
+  }
+
+  /* Clear all tables like public.etl_% */
+  const clearTables = async () => {
+    await pool.query('select etl_clear_tables();', async (err, result) => {
+      if (err) {
+        console.log(err)
```

Comment:
Some more messaging with a `console.error()` would be good for this and the ones below.  `console.error('Error executing query', err.stack)` comes to mind.

Reply:
Done